File: ios/RocChat/Crypto/GroupSessionManager.swift
import CryptoKit
import Foundation

// MARK: - Sender Key Models

struct SenderKeyState: Codable {
    var chainKey: Data       // 32-byte chain key for symmetric ratchet
    var iteration: Int       // Current chain step
    let senderIdHash: Data   // Identifies the sender
}

struct GroupCiphertext {
    let senderId: String
    let iteration: Int
    let ciphertext: Data     // AES-256-GCM encrypted
    let iv: Data
    let tag: Data
}

struct SenderKeyDistribution: Codable {
    let senderId: String
    let groupId: String
    let chainKey: String     // base64
    let iteration: Int
}

// MARK: - Group Session Manager (Sender Keys)

class GroupSessionManager {
    static let shared = GroupSessionManager()

    private let storagePrefix = "rocchat_group_keys_"
    private var ownKeys: [String: SenderKeyState] = [:]      // groupId → own sender key
    private var peerKeys: [String: SenderKeyState] = [:]      // groupId:senderId → peer sender key
    private var distributedTo: [String: Set<String>] = [:]    // groupId → set of userIds

    // MARK: - Key Generation

    func generateSenderKey() -> SenderKeyState {
        var random = Data(count: 32)
        random.withUnsafeMutableBytes { _ = SecRandomCopyBytes(kSecRandomDefault, 32, $0.baseAddress!) }
        let senderIdHash = SHA256.hash(data: random).withUnsafeBytes { Data($0) }
        return SenderKeyState(chainKey: random, iteration: 0, senderIdHash: senderIdHash)
    }

    // MARK: - Own Key Management

    func getOrCreateOwnKey(groupId: String) -> SenderKeyState {
        if let cached = ownKeys[groupId] { return cached }
        if let stored = loadKey(key: storagePrefix + "own_" + groupId) {
            ownKeys[groupId] = stored
            return stored
        }
        let key = generateSenderKey()
        ownKeys[groupId] = key
        saveKey(key, forKey: storagePrefix + "own_" + groupId)
        return key
    }

    // MARK: - Peer Key Management

    func getPeerKey(groupId: String, senderId: String) -> SenderKeyState? {
        let id = "\(groupId):\(senderId)"
        if let cached = peerKeys[id] { return cached }
        if let stored = loadKey(key: storagePrefix + "peer_" + id) {
            peerKeys[id] = stored
            return stored
        }
        return nil
    }

    func savePeerKey(groupId: String, senderId: String, key: SenderKeyState) {
        let id = "\(groupId):\(senderId)"
        peerKeys[id] = key
        saveKey(key, forKey: storagePrefix + "peer_" + id)
    }

    // MARK: - Distribution

    func createDistribution(senderId: String, groupId: String) -> SenderKeyDistribution {
        let key = getOrCreateOwnKey(groupId: groupId)
        return SenderKeyDistribution(
            senderId: senderId,
            groupId: groupId,
            chainKey: key.chainKey.base64EncodedString(),
            iteration: key.iteration
        )
    }

    func importDistribution(_ dist: SenderKeyDistribution) {
        guard let chainKey = Data(base64Encoded: dist.chainKey) else { return }
        let senderIdHash = SHA256.hash(data: chainKey).withUnsafeBytes { Data($0) }
        let key = SenderKeyState(chainKey: chainKey, iteration: dist.iteration, senderIdHash: senderIdHash)
        savePeerKey(groupId: dist.groupId, senderId: dist.senderId, key: key)
    }

    /// Distribute our sender key to all group members via pairwise channels.
    func ensureDistributed(
        groupId: String,
        members: [(userId: String, String)],
        myUserId: String
    ) async throws -> SenderKeyState {
        let senderKey = getOrCreateOwnKey(groupId: groupId)
        var distributed = distributedTo[groupId] ?? Set<String>()
        let otherMembers = members.filter { $0.userId != myUserId }
        let needsDistribution = otherMembers.filter { !distributed.contains($0.userId) }

        for member in needsDistribution {
            do {
                let dist = createDistribution(senderId: myUserId, groupId: groupId)
                let distJson = try JSONEncoder().encode(dist)
                let distStr = String(data: distJson, encoding: .utf8)!

                // Encrypt via pairwise Double Ratchet
                let envelope = try await SessionManager.shared.encryptMessage(
                    conversationId: groupId + ":" + member.userId,
                    recipientUserId: member.userId,
                    plaintext: distStr
                )

                // Send as sender_key_distribution message
                _ = try await APIClient.shared.postRaw("/messages/send", body: [
                    "conversation_id": groupId,
                    "ciphertext": envelope.ciphertext,
                    "iv": envelope.iv,
                    "ratchet_header": envelope.ratchetHeader,
                    "message_type": "sender_key_distribution",
                ])

                distributed.insert(member.userId)
            } catch {
                // Best effort — will retry on next send
            }
        }
        distributedTo[groupId] = distributed
        return senderKey
    }

    // MARK: - Encrypt

    private func advanceChain(_ key: inout SenderKeyState) -> Data {
        let info = "RocChat_SenderKey_v1".data(using: .utf8)!
        let derived = HKDF<SHA256>.deriveKey(
            inputKeyMaterial: SymmetricKey(data: key.chainKey),
            salt: Data(count: 32),
            info: info,
            outputByteCount: 64
        )
        let bytes = derived.withUnsafeBytes { Data($0) }
        key.chainKey = bytes[0..<32]
        key.iteration += 1
        return bytes[32..<64] // message key
    }

    func encrypt(groupId: String, plaintext: Data, myUserId: String) throws -> (ciphertext: String, ratchetHeader: String) {
        var key = getOrCreateOwnKey(groupId: groupId)
        let messageKey = advanceChain(&key)

        // AES-256-GCM encrypt
        let nonce = AES.GCM.Nonce()
        let sealed = try AES.GCM.seal(plaintext, using: SymmetricKey(data: messageKey), nonce: nonce)

        // Save updated key
        ownKeys[groupId] = key
        saveKey(key, forKey: storagePrefix + "own_" + groupId)

        // Wire format
        let ct = sealed.ciphertext + sealed.tag
        let header: [String: Any] = [
            "senderId": myUserId,
            "iteration": key.iteration,
            "groupEncrypted": true,
            "iv": Data(nonce).base64EncodedString(),
        ]
        let headerJSON = try JSONSerialization.data(withJSONObject: header)

        return (
            ciphertext: ct.base64EncodedString(),
            ratchetHeader: String(data: headerJSON, encoding: .utf8)!
        )
    }

    // MARK: - Decrypt

    func decrypt(groupId: String, senderId: String, ciphertextB64: String, ratchetHeader: String) throws -> Data {
        guard let headerData = ratchetHeader.data(using: .utf8),
              let header = try JSONSerialization.jsonObject(with: headerData) as? [String: Any],
              let iteration = header["iteration"] as? Int,
              let ivB64 = header["iv"] as? String,
              let ivData = Data(base64Encoded: ivB64),
              let ctData = Data(base64Encoded: ciphertextB64) else {
            throw GroupSessionError.invalidHeader
        }

        guard var peerKey = getPeerKey(groupId: groupId, senderId: senderId) else {
            throw GroupSessionError.noSenderKey
        }

        // Advance chain to match iteration
        while peerKey.iteration < iteration {
            _ = advanceChain(&peerKey)
        }

        let messageKey = advanceChain(&peerKey)
        savePeerKey(groupId: groupId, senderId: senderId, key: peerKey)

        // Split ciphertext + tag (last 16 bytes)
        guard ctData.count > 16 else { throw GroupSessionError.invalidCiphertext }
        let ciphertext = ctData[0..<(ctData.count - 16)]
        let tag = ctData[(ctData.count - 16)...]

        let nonce = try AES.GCM.Nonce(data: ivData)
        let box = try AES.GCM.SealedBox(nonce: nonce, ciphertext: ciphertext, tag: tag)
        return try AES.GCM.open(box, using: SymmetricKey(data: messageKey))
    }

    // MARK: - Handle incoming distribution

    func handleSenderKeyDistribution(
        groupId: String,
        senderId: String,
        ciphertext: String,
        iv: String,
        ratchetHeader: String
    ) throws {
        // Decrypt via pairwise Double Ratchet
        let decrypted = try SessionManager.shared.decryptMessage(
            conversationId: groupId + ":" + senderId,
            ciphertext: ciphertext,
            iv: iv,
            ratchetHeaderStr: ratchetHeader
        )
        guard let data = decrypted.data(using: .utf8),
              let dist = try? JSONDecoder().decode(SenderKeyDistribution.self, from: data) else { return }
        importDistribution(dist)
    }

    // MARK: - Key Rotation

    func rotateSenderKey(groupId: String) {
        let newKey = generateSenderKey()
        ownKeys[groupId] = newKey
        saveKey(newKey, forKey: storagePrefix + "own_" + groupId)
        distributedTo.removeValue(forKey: groupId)
    }

    // MARK: - Helpers

    static func isGroupEncrypted(ratchetHeader: String) -> Bool {
        guard let data = ratchetHeader.data(using: .utf8),
              let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else { return false }
        return json["groupEncrypted"] as? Bool == true
    }

    // MARK: - Persistence

    private func saveKey(_ key: SenderKeyState, forKey storageKey: String) {
        if let data = try? JSONEncoder().encode(key) {
            SecureStorage.shared.setData(data, forKey: storageKey)
        }
    }

    private func loadKey(key storageKey: String) -> SenderKeyState? {
        guard let data = SecureStorage.shared.getData(forKey: storageKey) else { return nil }
        return try? JSONDecoder().decode(SenderKeyState.self, from: data)
    }
}

enum GroupSessionError: Error {
    case noSenderKey, invalidHeader, invalidCiphertext
}

File: android/app/src/main/kotlin/com/rocchat/crypto/GroupSessionManager.kt
package com.rocchat.crypto

import android.content.Context
import android.util.Base64
import org.json.JSONObject
import java.security.MessageDigest
import java.security.SecureRandom
import javax.crypto.Cipher
import javax.crypto.Mac
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.SecretKeySpec

// MARK: - Models

data class SenderKeyState(
    var chainKey: ByteArray,
    var iteration: Int,
    val senderIdHash: ByteArray
) {
    fun toJSON(): JSONObject = JSONObject().apply {
        put("chainKey", Base64.encodeToString(chainKey, Base64.NO_WRAP))
        put("iteration", iteration)
        put("senderIdHash", Base64.encodeToString(senderIdHash, Base64.NO_WRAP))
    }
    companion object {
        fun fromJSON(json: JSONObject) = SenderKeyState(
            chainKey = Base64.decode(json.getString("chainKey"), Base64.NO_WRAP),
            iteration = json.getInt("iteration"),
            senderIdHash = Base64.decode(json.getString("senderIdHash"), Base64.NO_WRAP)
        )
    }
}

data class SenderKeyDistribution(
    val senderId: String,
    val groupId: String,
    val chainKey: String,
    val iteration: Int
) {
    fun toJSON(): JSONObject = JSONObject().apply {
        put("senderId", senderId)
        put("groupId", groupId)
        put("chainKey", chainKey)
        put("iteration", iteration)
    }
    companion object {
        fun fromJSON(json: JSONObject) = SenderKeyDistribution(
            senderId = json.getString("senderId"),
            groupId = json.getString("groupId"),
            chainKey = json.getString("chainKey"),
            iteration = json.getInt("iteration")
        )
    }
}

data class GroupEncryptResult(val ciphertext: String, val ratchetHeader: String)

// MARK: - Group Session Manager

class GroupSessionManager private constructor(private val context: Context) {
    companion object {
        @Volatile private var instance: GroupSessionManager? = null
        fun getInstance(context: Context): GroupSessionManager =
            instance ?: synchronized(this) {
                instance ?: GroupSessionManager(context.applicationContext).also { instance = it }
            }

        fun isGroupEncrypted(ratchetHeader: String): Boolean {
            return try {
                JSONObject(ratchetHeader).optBoolean("groupEncrypted", false)
            } catch (_: Exception) { false }
        }
    }

    private val prefs = context.getSharedPreferences("rocchat_group_keys", Context.MODE_PRIVATE)
    private val ownKeys = mutableMapOf<String, SenderKeyState>()
    private val peerKeys = mutableMapOf<String, SenderKeyState>()
    private val distributedTo = mutableMapOf<String, MutableSet<String>>()

    // MARK: - Key Generation

    fun generateSenderKey(): SenderKeyState {
        val random = ByteArray(32)
        SecureRandom().nextBytes(random)
        val hash = MessageDigest.getInstance("SHA-256").digest(random)
        return SenderKeyState(chainKey = random, iteration = 0, senderIdHash = hash)
    }

    // MARK: - Own Key Management

    fun getOrCreateOwnKey(groupId: String): SenderKeyState {
        ownKeys[groupId]?.let { return it }
        loadKey("own_$groupId")?.let {
            ownKeys[groupId] = it
            return it
        }
        val key = generateSenderKey()
        ownKeys[groupId] = key
        saveKey("own_$groupId", key)
        return key
    }

    // MARK: - Peer Key Management

    fun getPeerKey(groupId: String, senderId: String): SenderKeyState? {
        val id = "$groupId:$senderId"
        peerKeys[id]?.let { return it }
        loadKey("peer_$id")?.let {
            peerKeys[id] = it
            return it
        }
        return null
    }

    fun savePeerKey(groupId: String, senderId: String, key: SenderKeyState) {
        val id = "$groupId:$senderId"
        peerKeys[id] = key
        saveKey("peer_$id", key)
    }

    // MARK: - Distribution

    fun createDistribution(senderId: String, groupId: String): SenderKeyDistribution {
        val key = getOrCreateOwnKey(groupId)
        return SenderKeyDistribution(
            senderId = senderId,
            groupId = groupId,
            chainKey = Base64.encodeToString(key.chainKey, Base64.NO_WRAP),
            iteration = key.iteration
        )
    }

    fun importDistribution(dist: SenderKeyDistribution) {
        val chainKey = Base64.decode(dist.chainKey, Base64.NO_WRAP)
        val hash = MessageDigest.getInstance("SHA-256").digest(chainKey)
        val key = SenderKeyState(chainKey = chainKey, iteration = dist.iteration, senderIdHash = hash)
        savePeerKey(dist.groupId, dist.senderId, key)
    }

    suspend fun ensureDistributed(
        groupId: String,
        members: List<String>,
        myUserId: String,
    ): SenderKeyState {
        val senderKey = getOrCreateOwnKey(groupId)
        val distributed = distributedTo.getOrPut(groupId) { mutableSetOf() }
        val others = members.filter { it != myUserId }
        val needs = others.filter { it !in distributed }

        for (memberId in needs) {
            try {
                val dist = createDistribution(myUserId, groupId)
                val distStr = dist.toJSON().toString()

                val envelope = SessionManager.encryptMessage(
                    context, "$groupId:$memberId", memberId, distStr
                )

                val body = JSONObject().apply {
                    put("conversation_id", groupId)
                    put("ciphertext", envelope.ciphertext)
                    put("iv", envelope.iv)
                    put("ratchet_header", envelope.ratchetHeader)
                    put("message_type", "sender_key_distribution")
                }
                com.rocchat.network.APIClient.post("/messages/send", body)
                distributed.add(memberId)
            } catch (_: Exception) {
                // Best effort
            }
        }
        return senderKey
    }

    // MARK: - Chain Advance

    private fun advanceChain(key: SenderKeyState): ByteArray {
        val info = "RocChat_SenderKey_v1".toByteArray()
        val mac = Mac.getInstance("HmacSHA256")
        mac.init(SecretKeySpec(key.chainKey, "HmacSHA256"))
        mac.update(info)
        val derived = mac.doFinal()

        // Use HKDF-expand for 64 bytes
        val expand1 = Mac.getInstance("HmacSHA256").apply {
            init(SecretKeySpec(derived, "HmacSHA256"))
            update(info)
            update(byteArrayOf(1))
        }.doFinal()

        val expand2 = Mac.getInstance("HmacSHA256").apply {
            init(SecretKeySpec(derived, "HmacSHA256"))
            update(expand1)
            update(info)
            update(byteArrayOf(2))
        }.doFinal()

        key.chainKey = expand1.copyOf()
        key.iteration++
        return expand2 // message key
    }

    // MARK: - Encrypt

    fun encrypt(groupId: String, plaintext: ByteArray, myUserId: String): GroupEncryptResult {
        val key = getOrCreateOwnKey(groupId)
        val messageKey = advanceChain(key)
        ownKeys[groupId] = key
        saveKey("own_$groupId", key)

        // AES-256-GCM
        val iv = ByteArray(12)
        SecureRandom().nextBytes(iv)
        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        cipher.init(Cipher.ENCRYPT_MODE, SecretKeySpec(messageKey, "AES"), GCMParameterSpec(128, iv))
        val ct = cipher.doFinal(plaintext)

        val header = JSONObject().apply {
            put("senderId", myUserId)
            put("iteration", key.iteration)
            put("groupEncrypted", true)
            put("iv", Base64.encodeToString(iv, Base64.NO_WRAP))
        }

        return GroupEncryptResult(
            ciphertext = Base64.encodeToString(ct, Base64.NO_WRAP),
            ratchetHeader = header.toString()
        )
    }

    // MARK: - Decrypt

    fun decrypt(groupId: String, senderId: String, ciphertextB64: String, ratchetHeader: String): ByteArray {
        val header = JSONObject(ratchetHeader)
        val iteration = header.getInt("iteration")
        val ivB64 = header.getString("iv")
        val iv = Base64.decode(ivB64, Base64.NO_WRAP)
        val ctData = Base64.decode(ciphertextB64, Base64.NO_WRAP)

        val peerKey = getPeerKey(groupId, senderId) ?: throw IllegalStateException("No sender key for $senderId in group $groupId")

        // Advance to match iteration
        while (peerKey.iteration < iteration) {
            advanceChain(peerKey)
        }

        val messageKey = advanceChain(peerKey)
        savePeerKey(groupId, senderId, peerKey)

        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        cipher.init(Cipher.DECRYPT_MODE, SecretKeySpec(messageKey, "AES"), GCMParameterSpec(128, iv))
        return cipher.doFinal(ctData)
    }

    // MARK: - Handle incoming distribution

    fun handleSenderKeyDistribution(
        groupId: String,
        senderId: String,
        ciphertext: String,
        iv: String,
        ratchetHeader: String,
    ) {
        val decrypted = SessionManager.decryptMessage(context, "$groupId:$senderId", ciphertext, iv, ratchetHeader)
        val dist = SenderKeyDistribution.fromJSON(JSONObject(decrypted))
        importDistribution(dist)
    }

    // MARK: - Key Rotation

    fun rotateSenderKey(groupId: String) {
        val newKey = generateSenderKey()
        ownKeys[groupId] = newKey
        saveKey("own_$groupId", newKey)
        distributedTo.remove(groupId)
    }

    // MARK: - Persistence

    private fun saveKey(key: String, state: SenderKeyState) {
        prefs.edit().putString(key, state.toJSON().toString()).apply()
    }

    private fun loadKey(key: String): SenderKeyState? {
        val json = prefs.getString(key, null) ?: return null
        return try { SenderKeyState.fromJSON(JSONObject(json)) } catch (_: Exception) { null }
    }
}
